### PR TITLE
Modify session creation to work with AssumeRoleWithWebIdentity

### PR DESCRIPTION
**Pull request by throwanexception**
Howdy! 

Creating the config struct via defaults.Get() has the unfortunate side effect of precluding other authentication behaviors including AssumeRoleWithWebIdentity - which allows folks who use IAM Roles with Service Accounts (AWS / EKS feature) to map Service Accounts to IAM Roles and use it as a finer grained authentication mechanism than ec2 instance roles.

I was unsure what behavior was being relied on with defaults.Get() so this is probably imperfect! I'm happy to iterate to get it accepted (or feel free to alter yourself!)

### CODE REVIEW

**Comment by jniebuhr** [comment]
Hi @throwanexception, thanks for your pull request. The goal with using the default config was getting the default provider chain. (You can see the order under https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/configuring-sdk.html#specifying-credentials) By default you're able to authenticate with environment variables, a profile file or an ec2 instance role.

We should definitely add the use case you're describing. From what I've read, it will provide the environment variables AWS_WEB_IDENTITY_TOKEN_FILE and AWS_ROLE_ARN. So we should react to those. 
What I would suggest is implementing a custom provider that does this (https://docs.aws.amazon.com/sdk-for-go/api/aws/credentials/) and adding it to the chain. The basics of assuming a role via web identity is already implemented in https://docs.aws.amazon.com/sdk-for-go/api/aws/credentials/stscreds/#NewWebIdentityRoleProvider so you'll just have to implement reacting to the environment variables and reading the token.

**Comment by throwanexception** [comment]
`envConfigLoad()` is supposed to be called as part of `session.NewSession()` and should already pick up on a WebIdentityTokenFilePath
https://github.com/aws/aws-sdk-go/blob/65c0691f756ce0f4436b98b5d5690872894454ba/aws/session/env_config.go#L266

My understanding from reading session.NewSession() is that it should rely on whether AWS_SDK_LOAD_CONFIG is set, and automatically load the shared environment configuration to setup credentials appropriately.

It appears that defaults.Get(), as it's merged into NewSession(), obviates this behavior in some way:
```
{"level":"error","ts":1612392759.2262459,"logger":"controllers.CertificateRequest","msg":"failed to request certificate from PCA","certificaterequest":"cert-manager/rsa-cert-2048-5lctk","error":"NoCredentialProviders: no valid providers in chain. Deprecated.\n\tFor verbose messaging see aws.Config.CredentialsChainVerboseErrors","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/pkg/mod/github.com/go-logr/zapr@v0.2.0/zapr.go:132\ngithub.com/jniebuhr/aws-pca-issuer/pkg/controllers.
...
```
I'm not 100% sure of the cause - I don't understand aws-sdk-go well enough - but I think it might be https://github.com/aws/aws-sdk-go/blob/1f88422051125aee6a38362ffe14a843a9e095c1/aws/defaults/defaults.go#L107 the handlers that are created here taking precedence.

I suppose that leads me to the actual question - I understand what you're asking, but why implement this behavior separately when simply calling session.NewSession( ) without passing defaults.get() already provides the expected behavior?

I'm unsure if the fact that the defaults package also suggests:
```
Generally this package shouldn't be used directly, but session.Session instead. This package is useful when you need to reset the defaults of a session or service client to the SDK defaults before setting additional parameters.
```
https://docs.aws.amazon.com/sdk-for-go/api/aws/defaults
Implies that defaults wasn't intended to be used to create a config object to be passed at session creation?

https://docs.aws.amazon.com/sdk-for-go/api/aws/session/#NewSession
```
// NewSession returns a new Session created from SDK defaults, config files,
// environment, and user provided config files. Once the Session is created
// it can be mutated to modify the Config or Handlers. The Session is safe to
// be read concurrently, but it should not be written to concurrently.
```

I guess my concern is that how it's currently creating a session is breaking the expected behavior of the client in the first place.

**Comment by jniebuhr** [comment]
Hi, thanks for the clarification. I didn’t understand that simply calling it without parameters produced that behavior. I will test it today and merge it afterwards.

File: pkg/controllers/genericissuer_controller.go [hunk 1]
```diff
@@ -62,12 +61,21 @@ func (r *GenericIssuerReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 		return ctrl.Result{}, err
 	}
 
-	config := defaults.Get()
-
+	var region *string
```
Comment (jniebuhr) [inline pkg/controllers/genericissuer_controller.go:64]:
```suggestion
        config := aws.Config{}
        if spec.Region != "" {
		config.Region = aws.String(spec.Region)
	}
```

File: pkg/controllers/genericissuer_controller.go [hunk 2]
```diff
@@ -97,10 +105,13 @@ func (r *GenericIssuerReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 			return ctrl.Result{}, err
 		}
 
-		config.Config.Credentials = credentials.NewStaticCredentials(string(accessKey), string(secretKey), "")
+		clientCredentials = credentials.NewStaticCredentials(string(accessKey), string(secretKey), "")
```
Comment (jniebuhr) [inline pkg/controllers/genericissuer_controller.go:108]:
```suggestion
		config.clientCredentials = credentials.NewStaticCredentials(string(accessKey), string(secretKey), "")
```

File: pkg/controllers/genericissuer_controller.go [hunk 3]
```diff
@@ -97,10 +105,13 @@ func (r *GenericIssuerReconciler) Reconcile(ctx context.Context, req ctrl.Reques
...
 	}
 
-	sess, err := session.NewSession(config.Config)
+	sess, err := session.NewSession(&aws.Config{
```
Comment (jniebuhr) [inline pkg/controllers/genericissuer_controller.go:111]:
```suggestion
	sess, err := session.NewSession(&config)
```

**Review by jniebuhr — CHANGES_REQUESTED** [review]
I've tested the functionality, can you make these changes so we're touching the config as little as possible?

**Comment by jniebuhr** [comment]
Closed with commit 1375fd7c6c4ca774348e3957ed049988342ae35e

**Comment by jniebuhr** [comment]
I have released v0.1.2 with the fix

**Pull request closed without merging**